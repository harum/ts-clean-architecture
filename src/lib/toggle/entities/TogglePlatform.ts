import Toggle from './Toggle';

export default class TogglePlatform extends Toggle {
  platform: string;

  constructor(platform = '') {
    super(false);
    this.platform = platform;
  }

  buildFromHash(platformHash: { platform: string; active: boolean }) {
    this.platform = platformHash.platform;
    if (platformHash.active) {
      this.on();
    } else {
      this.off();
    }
  }

  toString(): string {
    return JSON.stringify(this.toHash());
  }

  toHash(): { platform: string; active: boolean } {
    return {
      platform: this.platform,
      active: this.active,
    };
  }
}
