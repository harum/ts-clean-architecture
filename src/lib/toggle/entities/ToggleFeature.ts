import ToggleAble from './ToggleAble';
import TogglePlatform from './TogglePlatform';
import ToggleAbleFeature from './ToggleAbleFeature';

export default class ToggleFeature implements ToggleAbleFeature {
  #platforms = new Map<string, TogglePlatform>();
  feature: string;

  constructor(feature: string) {
    this.feature = feature;
  }

  get platforms(): Map<string, TogglePlatform> {
    return this.#platforms;
  }

  get platformsString(): string {
    const results = [];
    for (const [_, platformToggle] of this.#platforms.entries()) {
      results.push(platformToggle.toString());
    }

    return results.join(', ');
  }

  active(platform: string): boolean {
    const togglePlatform = this.#platforms.get(platform);
    return (togglePlatform && togglePlatform.active) || false;
  }

  toggle(platform: string): void {
    const togglePlatform = this.#platforms.get(platform);
    if (togglePlatform) {
      togglePlatform.toggle();
    }
  }

  on(platform: string): void {
    const togglePlatform = this.#platforms.get(platform);
    if (togglePlatform) {
      togglePlatform.on();
    }
  }

  off(platform: string): void {
    const togglePlatform = this.#platforms.get(platform);
    if (togglePlatform) {
      togglePlatform.off();
    }
  }

  addPlatform(togglePlatform: TogglePlatform) {
    this.#platforms.set(togglePlatform.platform, togglePlatform);
  }

  deletePlatform(platform: string) {
    this.#platforms.delete(platform);
  }

  toString() {
    return (
      'ToggleFeature{' +
      `feature: ${this.feature}` +
      `, platforms: ${this.platformsString}` +
      '}'
    );
  }
}
