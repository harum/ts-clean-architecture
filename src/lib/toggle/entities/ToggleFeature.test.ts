import ToggleFeature from './ToggleFeature';
import TogglePlatform from './TogglePlatform';

describe('ToggleFeature', () => {
  let toggleFeature: ToggleFeature;

  beforeEach(() => {
    toggleFeature = new ToggleFeature('test');
  });

  describe('When toggle feature is initialized', () => {
    it('should be able to have initial value', () => {
      expect(toggleFeature.toString()).toMatchSnapshot();
    });
  });

  describe('When targeted platform in not registered', () => {
    it('should stay off when toggle on the platform', () => {
      toggleFeature.on('mobile');
      expect(toggleFeature.active('mobile')).toBe(false);
    });

    it('should stay off when toggle off a platform', () => {
      toggleFeature.on('mobile');
      toggleFeature.off('mobile');
      expect(toggleFeature.active('mobile')).toBe(false);
    });

    it('should stay off when toggle a platform', () => {
      toggleFeature.toggle('mobile');
      expect(toggleFeature.active('mobile')).toBe(false);
    });
  });

  describe('When targeted platform is registered', () => {
    beforeEach(() => {
      toggleFeature.addPlatform(new TogglePlatform('mobile'));
    });

    it('should be able to toggle on the platform', () => {
      toggleFeature.on('mobile');
      expect(toggleFeature.active('mobile')).toBe(true);
    });

    it('should be able to toggle off the platform', () => {
      toggleFeature.on('mobile');
      toggleFeature.off('mobile');
      expect(toggleFeature.active('mobile')).toBe(false);
    });

    it('should be able to toggle the platform', () => {
      toggleFeature.toggle('mobile');
      expect(toggleFeature.active('mobile')).toBe(true);

      toggleFeature.toggle('mobile');
      expect(toggleFeature.active('mobile')).toBe(false);
    });
  });

  describe('When platform is more then one', () => {
    let platform1: TogglePlatform;
    let platform2: TogglePlatform;

    beforeEach(() => {
      platform1 = new TogglePlatform('desktop');
      platform2 = new TogglePlatform('mobile');

      toggleFeature.addPlatform(platform1);
      toggleFeature.addPlatform(platform2);
    });

    it('should be able to have correct platforms', () => {
      expect(toggleFeature.toString()).toMatchSnapshot();
      expect(toggleFeature.platforms.get('desktop')).toEqual(platform1);
      expect(toggleFeature.platforms.get('mobile')).toEqual(platform2);
    });

    it('should be able to delete platform', () => {
      toggleFeature.deletePlatform('mobile');
      expect(toggleFeature.toString()).toMatchSnapshot();
    });
  });
});
