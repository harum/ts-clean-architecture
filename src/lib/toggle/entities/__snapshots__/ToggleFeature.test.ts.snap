// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleFeature When platform is more then one should be able to delete platform 1`] = `"ToggleFeature{feature: test, platforms: TogglePlatform{platform: desktop, active: false}}"`;

exports[`ToggleFeature When platform is more then one should be able to have correct platforms 1`] = `"ToggleFeature{feature: test, platforms: TogglePlatform{platform: desktop, active: false}, TogglePlatform{platform: mobile, active: false}}"`;

exports[`ToggleFeature When toggle feature is initialized should be able to have initial value 1`] = `"ToggleFeature{feature: test, platforms: }"`;
